### PR TITLE
Migrate to use the node GRPC API V2

### PR DESCRIPTION
**Pull request by vikt0r0**
### Purpose
The client currently leverages the v1 GRPC API exposed by the node. This feature aims to migrate the client to use the v2 GRPC API while simultaneously preserving the API of the client. 

Using the node GRPC API v2 has a number of advantages, including lower request latency, higher request throughput (at the node side), the ability to leverage streaming APIs as well as having a typed messages. The current version of the API also allows for more granular error handling as it leverages GRPC status codes.

### Changes
- [x] Add bootstrapping to generate code from the V2 API protocol buffer schema (in `Setup.hs`).
- [x] Migrate GRPC call helpers to use the V2 GRPC API:
- [x] Add functional "inverse" to the `ToProto` class (i.e. `FromProto`), allowing for conversion between datatypes generated from protocol buffer schemas and their corresponding datatype in (e.g. those in `Concordium.Types.*`, `Concordium.Common`) where applicable. These are not necessarily one-to-one, so checks should be added accordingly.
- [x] (Maybe) refactor `ToProto` and `FromProto` class and instance declarations to `concordium-base`.
- [x] Implement V2 GRPC API calls.
- [x] Rewrite/refactor client functions and helpers to use V2 GRPC API
- [x] Rewrite middleware to use the V2 GRPC API.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: src/Concordium/Client/GRPC2.hs [hunk 1]
```diff
@@ -0,0 +1,3372 @@
...
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- |Part of the implementation of the GRPC2 interface.
```
Comment (abizjak) [inline src/Concordium/Client/GRPC2.hs:11]: Please elaborate on this.

**Comment by abizjak** [comment]
Additionally, please change the version of the client by bumping the minor version.

And add to the changelog that the defaults have changed wrt the node API.

- [ ] In the #Usage section of the README explain that you need to connect to V2 API.
- [ ] Remove the generation of the V1 API in the build script.

File: src/Concordium/Client/Output.hs [hunk 2]
```diff
@@ -985,6 +996,33 @@ printConsensusStatus r =
        , printf "Current era genesis block:   %s" (show $ Queries.csCurrentEraGenesisBlock r)
        , printf "Current era genesis time:    %s" (show $ Queries.csCurrentEraGenesisTime r)]
 
+-- |Print Birk parameters from a @BlockBirkParameters@.
+printQueryBirkParameters :: Bool -> Queries.BlockBirkParameters -> Map.Map IDTypes.AccountAddress Text -> Printer
+printQueryBirkParameters includeBakers r addrmap = do
+  tell [ printf "Election nonce:      %s" (show $ Queries.bbpElectionNonce r)
+      ] --, printf "Election difficulty: %f" (Types.electionDifficulty $ bprElectionDifficulty r) ]
```
[the remark this answers is not in the document]
Comment (vikt0r0) [reply src/Concordium/Client/Output.hs:1003]: This was not in the original output, so I originally  kept it this way to be consistent. `BirkParametersResult` does not carry this data for some reason. I added it now 👍 

(It is also printed in `consensus show-chain-parameters` as well)

File: src/Concordium/Client/Output.hs [hunk 3]
```diff
@@ -985,6 +996,33 @@ printConsensusStatus r =
...
+  when includeBakers $
+    case Vec.toList $ Queries.bbpBakers r of
+      [] ->
+         tell [ "Bakers:              " ++ showNone ]
+      bakers -> do
+        tell [ "Bakers:"
+             , printf "                             Account                       Lottery power  Account Name"
+             , printf "        ------------------------------------------------------------------------------" ]
```
Comment (Bargsteen) [inline src/Concordium/Client/Output.hs:1011]: You can remove these `printf` calls and just have the strings.

File: src/Concordium/Client/Types/TransactionStatus.hs [hunk 4]
```diff
@@ -25,6 +27,31 @@ data TransactionStatusResult' a = TransactionStatusResult
 
 type TransactionStatusResult = TransactionStatusResult' ValidResult
 
+-- |Convert a @TransactionStatus@ instance into a @TransactionStatusResult@ instance.
+-- Returns @Left@ wrapping an error message if a transaction summary was @Nothing@ if
+-- the input is either @Committed@ or @Finalized@, or a @Right@ wrapping the result
+-- otherwise.
```
Comment (Bargsteen) [inline src/Concordium/Client/Types/TransactionStatus.hs:33]: This sentence is a bit convoluted and ambiguous. Can you make it clearer?

File: src/Concordium/Client/Runner.hs [hunk 5]
```diff
@@ -3438,95 +3572,253 @@ processIdentityShowCmd :: IdentityShowCmd -> Backend -> IO ()
...
+    -- The boolean indicates whether to recurse on the ancestor
+    -- of the block. The recursion bottoms out when the genesis
+    -- block is reached.
+    printBlockInfos :: Bool -> BlockHashInput -> ClientMonad IO ()
+    printBlockInfos recurse bh = do
+      bi <- getResponseValueOrDie =<< getBlockInfoV2 bh
+      printJSONValues $ toJSON bi
+      -- Recurse if were instructed to and if we are not at the genesis block.
+      when (recurse && Queries.biBlockHeight bi /= 0)
+        (printBlockInfos recurse $ Given (Queries.biBlockParent bi))
```
Comment (Bargsteen) [inline src/Concordium/Client/Runner.hs:3820]: Consider moving these `print*` functions to `Output.hs`.

**Review by Bargsteen — COMMENTED** [review]
LGTM if Ales' demands will be met ;)

File: src/Concordium/Client/Runner.hs (hunk 5 above)
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3820]: I could do that, but I am not entirely sure under which heading they would belong (cfr. the organization of that file).

File: ChangeLog.md [hunk 6]
```diff
@@ -2,6 +2,36 @@
 
 ## Unreleased
 
+- The client now uses the V2 GRPC API exposed by the node. This introduces some breaking
+  changes:
+    - Since the node serves the V2 GRPC API at port 20000 by default, the default value of
+      the `--grpc-port` option has been updated to reflect this.
```
Comment (abizjak) [inline ChangeLog.md:8]: We should remove --grpc-token and mention it here.

File: src/Concordium/Client/Commands.hs [hunk 7]
```diff
@@ -540,7 +540,7 @@ portParser =
     auto
```
Comment (abizjak) [inline src/Concordium/Client/Commands.hs:540]: the authentication token parser is still used. Please remove it.

File: ChangeLog.md (hunk 6 above)
Comment (vikt0r0) [reply ChangeLog.md:8]: I assume by this you refer to `--grpc-authentication-token`?

**Review by Bargsteen — APPROVED** [review]
Great work 😊 LGTM

**Comment by vikt0r0** [comment]
> Great work 😊 LGTM

Thank you Kasper! And thanks for your useful review comments :-)

**Review by abizjak — APPROVED** [review]

**Pull request merged**